### PR TITLE
Fix constant folding bug with sym size tensor

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118411

When there was a constant folded SymInt which was used to construct a then constant folding tensor, we had previously used tried to use the sympy symbol which would error (should take in SymInt not symbol). 

Fix by recording the observed size during constant folding.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/fx_passes/joint_graph.py [hunk 1]
```diff
@@ -238,7 +242,7 @@ def constant_fold_uniform_value(gm: torch.fx.GraphModule):
             # zeros, and ones just get traced into full, so we insert those
             new_node = graph.call_function(
                 aten.full.default,
-                args=(list(fake_tensor.shape), value),
+                args=(list(node_replacements_shapes[node]), value),
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/fx_passes/joint_graph.py:245]: We're taking the Size from the runtime tensor which will not SymInts. I can make it `List[int]` instead for better typing.

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_inductor/fx_passes/joint_graph.py]: This still seems suspicious? Suppose previously the representation of the tensor in the graph had dynamic shape (e.g., was expressed in terms of s0). After this change, you will specialize the tensor. To make matters worse, it's not clear to me that you have installed the right guard!

What I was *expecting* to see was to rewrite the transformation to preserve the input node structure for sizes, rewriting only other parts of the constructor?

**Review by ezyang — APPROVED** [review]
I'd like a comment elaborating on what we discussed in WC

**Pull request closed without merging**